src/python/pants/engine/internals/rule_visitor_test.py

```diff
@@ -142,6 +142,37 @@ async def rule3() -> int:
     assert_awaitables(rule3, [(int, []), (int, int), (int, [])])
 
 
+@rule
+async def recursive_rule(arg: int) -> int:
```

Comment:
Note that the rules used by the tests must be at the module level, so that they're pushed onto the stack in the _TypeStack ctor. If we ever need to support call-by-name in a @rule nested inside some inner scope we'll have to jump through more hoops. But that doesn't seem like an important case to support right now.